py-polars/polars/interchange/from_dataframe.py

```diff
@@ -37,41 +44,277 @@ def from_dataframe(df: SupportsInterchange, *, allow_copy: bool = True) -> DataF
-            " to a Polars dataframe"
+def _from_dataframe(df: InterchangeDataFrame, *, allow_copy: bool = True) -> DataFrame:
+    chunks = []
+    for chunk in df.get_chunks():
+        polars_chunk = _protocol_df_chunk_to_polars(chunk, allow_copy=allow_copy)
+        chunks.append(polars_chunk)
+
+    # Handle implementations that incorrectly yield no chunks for an empty dataframe
+    if not chunks:
+        polars_chunk = _protocol_df_chunk_to_polars(df, allow_copy=allow_copy)
+        chunks.append(polars_chunk)
+
+    return F.concat(chunks, rechunk=False)
+
+
+def _protocol_df_chunk_to_polars(
+    df: InterchangeDataFrame, *, allow_copy: bool = True
+) -> DataFrame:
+    columns = []
+    for column, name in zip(df.get_columns(), df.column_names()):
+        dtype = dtype_to_polars_dtype(column.dtype)
+        if dtype == String:
+            s = _string_column_to_series(column, allow_copy=allow_copy)
+        elif dtype == Enum:
+            s = _categorical_column_to_series(column, allow_copy=allow_copy)
+        else:
+            s = _column_to_series(column, dtype, allow_copy=allow_copy)
+        columns.append(s.alias(name))
+
+    return pl.DataFrame(columns)
+
+
+def _column_to_series(
+    column: Column, dtype: PolarsDataType, *, allow_copy: bool = True
+) -> Series:
+    buffers = column.get_buffers()
+    offset = column.offset
+
+    data_buffer = _construct_data_buffer(
+        *buffers["data"], column.size(), offset, allow_copy=allow_copy
+    )
+    validity_buffer = _construct_validity_buffer(
+        buffers["validity"], column, dtype, data_buffer, offset, allow_copy=allow_copy
+    )
+    return pl.Series._from_buffers(dtype, data=data_buffer, validity=validity_buffer)
+
+
+def _string_column_to_series(column: Column, *, allow_copy: bool = True) -> Series:
+    buffers = column.get_buffers()
+    offset = column.offset
+
+    offsets_buffer_info = buffers["offsets"]
+    if offsets_buffer_info is None:
+        msg = "cannot create String column without an offsets buffer"
+        raise RuntimeError(msg)
+    offsets_buffer: Series = _construct_offsets_buffer(
+        *offsets_buffer_info, allow_copy=allow_copy
+    )
+
+    buffer, dtype = buffers["data"]
+    data_buffer = _construct_data_buffer(
+        buffer, dtype, buffer.bufsize, offset=0, allow_copy=allow_copy
+    )
+
+    # First construct a Series without a validity buffer
+    # to allow constructing the validity buffer from a sentinel value
+    data_buffers: list[Series] = [data_buffer, offsets_buffer]
+    data = pl.Series._from_buffers(String, data=data_buffers, validity=None)
+    data = data[offset:]
+
+    # Add the validity buffer if present
+    validity_buffer = _construct_validity_buffer(
+        buffers["validity"], column, String, data, offset, allow_copy=allow_copy
+    )
+    if validity_buffer is not None:
+        data = pl.Series._from_buffers(
+            String, data=data_buffers, validity=validity_buffer
         )
 
```

Comment:
is this right?
```python
df = pd.DataFrame({'a': ['a', 'b', None, 'c', None, 'd']}, dtype='string[pyarrow]')
dfpa = pa.Table.from_pandas(df)
print(pl.from_dataframe(dfpa[1:]))
```

gives

```
Traceback (most recent call last):
  File "/home/marcogorelli/polars-dev/py-polars/t.py", line 12, in <module>
    print(pl.from_dataframe(dfpa[1:]))
  File "/home/marcogorelli/polars-dev/py-polars/polars/convert.py", line 770, in from_dataframe
    return from_dataframe(df, allow_copy=allow_copy)
  File "/home/marcogorelli/polars-dev/py-polars/polars/interchange/from_dataframe.py", line 47, in from_dataframe
    return _from_dataframe(
  File "/home/marcogorelli/polars-dev/py-polars/polars/interchange/from_dataframe.py", line 56, in _from_dataframe
    polars_chunk = _protocol_df_chunk_to_polars(chunk, allow_copy=allow_copy)
  File "/home/marcogorelli/polars-dev/py-polars/polars/interchange/from_dataframe.py", line 74, in _protocol_df_chunk_to_polars
    s = _string_column_to_series(column, allow_copy=allow_copy)
  File "/home/marcogorelli/polars-dev/py-polars/polars/interchange/from_dataframe.py", line 127, in _string_column_to_series
    data = pl.Series._from_buffers(
  File "/home/marcogorelli/polars-dev/py-polars/polars/series/series.py", line 492, in _from_buffers
    return self._from_pyseries(PySeries._from_buffers(dtype, data, validity))
pyo3_runtime.PanicException: called `Result::unwrap()` on an `Err` value: ComputeError(ErrString("validity mask length must match the number of values"))
```

Reply:
It's not! I tested chunked strings but not chunked strings with nulls... update incoming.

Reply:
thanks! I remembered a similar one from pandas so it was one of the first things I checked https://github.com/pandas-dev/pandas/issues/52845

I'll take a close look at the rest this week